File: credit.c
#include <stdio.h>
#include <cs50.h>

long checkthesum (long x);

int main (void)
{
    long answer;
    int checksum;

    // Prompt user for credit card number until last digit of checksum is 0
    do
    {
        answer = get_long ("Number: ");
    }
    while (answer < 0);

    long z = checkthesum(answer);

    if (z == 0)
    {
        if ((answer / 10000000000000) % 100 == 34 || (answer / 10000000000000) % 100 == 37)
        {
            printf("AMEX\n");
        }
        else if ((answer / 100000000000000) % 100 == 51 || (answer / 100000000000000) % 100 == 52 || (answer / 100000000000000) % 100 == 53 || (answer / 100000000000000) % 100 == 54 || (answer / 100000000000000) % 100 == 55)
        {
            printf("MASTERCARD\n");
        }
        else if ((answer / 1000000000000000) % 10 == 4 || (answer / 1000000000000) % 10 == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

// Define function that returns the last digit of the checksum
long checkthesum (long x)
{
    int checksum;
    int n = 0;
    int m = 0;
    int o = 0;

    while (x != 0)
    {
        long lastdigit = x % 10;
        long newanswer = (x - lastdigit) / 10;
        long secondtolastdigit = newanswer % 10;
        long newnewanswer = (newanswer - secondtolastdigit) / 10;
        n = (2 * secondtolastdigit);
        o = ((n % 10) + (n - (n % 10))/10) + o;
        m = lastdigit + m;
        x = newnewanswer;
    }

    // Add numbers and find last digit
    int p = m + o;
    checksum = p % 10;
    return checksum;
}

File: credit.py
from cs50 import get_string
import re
import sys

# Get user input and verify that it could be a legit number using regular expressions
while True:
    card_number = get_string("Number: ")
    regexp = re.compile(r'(5[1-5]|3[47]|4)\d{12,15}')
    if not regexp.search(card_number):
        print("INVALID")
        sys.exit(0)
    if regexp.search(card_number):
        break

# Create two lists starting with the second_to_last_digit and starting with the last_digit
last_digit_list = []
second_to_last_digit_list = []

# Append the correct items to the second_to_last_digit_list
for i in range(2, len(card_number), 2):
    second_to_last_digit_list.append(card_number[len(card_number) - i])
if len(card_number) % 2 == 0:
    second_to_last_digit_list.append(card_number[0])

# Append the correct items to the last_digit_list
for j in range(1, len(card_number), 2):
    last_digit_list.append(card_number[len(card_number) - j])
if not len(card_number) % 2 == 0:
    last_digit_list.append(card_number[0])

# Find the checksum value to see if it ends in 0

# Initiate the sums that will be added to make the checksum
sum1 = 0
sum2 = 0

# Calculate the sums
for x in range(len(last_digit_list)):
    last_digit = int(last_digit_list[x])
    sum1 += last_digit

for y in range(len(second_to_last_digit_list)):
    second_to_last_digit = int(second_to_last_digit_list[y]) * 2
    second_to_last_digit = ((second_to_last_digit - second_to_last_digit % 10) / 10) + second_to_last_digit % 10
    sum2 += second_to_last_digit

# Calculate the checksum
checksum = (sum1 + sum2) % 10

# Print the corresponding correct card or Invalid if the checksum != 0
if checksum == 0:
    if int(card_number[0]) == 3:
        print("AMEX")
    elif int(card_number[0]) == 4:
        print("VISA")
    elif int(card_number[0]) == 5:
        print("MASTERCARD")
    else:
        print("INVALID")
else:
    print("INVALID")

"""
for index, digit in enumerate(last_digit_list):
    last_digit = digit
    last_index = index
"""

"""int_card_number = int(card_number)
last_digit = int_card_number % 10
second_to_last_digit = (int_card_number - last_digit) % 10

print(last_digit)
print(second_to_last_digit)"""

# Find Checksum, convert the user input into an int
int_card_number = int(card_number)

real_numbers = '''
378282246310005,
371449635398431,
5555555555554444,
5105105105105100,
4111111111111111,
4012888888881881,
1234567890
'''